File: host/host-storage/test-user-1770963792-build-c7ab005de96e-2e779cb4/src/polymarket_utils.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"
)

const (
	gammaMarketsBySlug = "https://gamma-api.polymarket.com/markets?slug="
	gammaUserAgent     = "Mozilla/5.0 (compatible; CodexBot/1.0)"
)

var (
	monthNames = []string{
		"",
		"january",
		"february",
		"march",
		"april",
		"may",
		"june",
		"july",
		"august",
		"september",
		"october",
		"november",
		"december",
	}
	slugTimeRE = regexp.MustCompile(`^(?P<prefix>.+)-(?P<month>january|february|march|april|may|june|july|august|september|october|november|december)-(?P<day>\d{1,2})-(?P<hour>\d{1,2})(?P<ampm>am|pm)-et$`)
)

func normalizeSlug(value string) string {
	marker := "polymarket.com/event/"
	if idx := strings.Index(value, marker); idx >= 0 {
		return strings.Trim(strings.TrimPrefix(value[idx:], marker), "/")
	}
	return value
}

func fetchMarketBySlug(slug string) (map[string]any, error) {
	slug = normalizeSlug(slug)
	u, err := url.Parse(gammaMarketsBySlug + url.QueryEscape(slug))
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequest(http.MethodGet, u.String(), nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("User-Agent", gammaUserAgent)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	payload, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	var data any
	if err := json.Unmarshal(payload, &data); err != nil {
		return nil, err
	}

	markets := []any{}
	switch v := data.(type) {
	case map[string]any:
		if raw, ok := v["markets"]; ok {
			markets = toSlice(raw)
		} else if raw, ok := v["data"]; ok {
			markets = toSlice(raw)
		}
	case []any:
		markets = v
	}
	if len(markets) == 0 {
		return nil, fmt.Errorf("no market found for slug: %s", slug)
	}
	row, ok := markets[0].(map[string]any)
	if !ok {
		return nil, fmt.Errorf("unexpected market payload")
	}
	return row, nil
}

func inferSlugPrefix(slug string) string {
	slug = normalizeSlug(slug)
	match := slugTimeRE.FindStringSubmatch(slug)
	if match == nil {
		return ""
	}
	for i, name := range slugTimeRE.SubexpNames() {
		if name == "prefix" && i < len(match) {
			return match[i]
		}
	}
	return ""
}

func buildSlug(prefix string, whenET time.Time) string {
	month := monthNames[int(whenET.Month())]
	day := whenET.Day()
	hour24 := whenET.Hour()
	hour12 := hour24 % 12
	if hour12 == 0 {
		hour12 = 12
	}
	ampm := "am"
	if hour24 >= 12 {
		ampm = "pm"
	}
	return fmt.Sprintf("%s-%s-%d-%d%s-et", prefix, month, day, hour12, ampm)
}

func parseISODatetime(value string) *time.Time {
	if value == "" {
		return nil
	}
	if strings.HasSuffix(value, "Z") {
		value = strings.TrimSuffix(value, "Z") + "+00:00"
	}
	layouts := []string{time.RFC3339Nano, time.RFC3339, "2006-01-02T15:04:05-07:00", "2006-01-02T15:04:05"}
	for _, layout := range layouts {
		if t, err := time.Parse(layout, value); err == nil {
			return &t
		}
	}
	return nil
}

func isOpenMarket(market map[string]any, nowUTC time.Time) bool {
	if v, ok := market["closed"]; ok {
		if b, ok := v.(bool); ok && b {
			return false
		}
	}
	if start := parseISODatetime(toString(market["startDate"])); start != nil {
		if nowUTC.Before(start.UTC()) {
			return false
		}
	}
	if end := parseISODatetime(toString(market["endDate"])); end != nil {
		if end.UTC().Before(nowUTC) {
			return false
		}
	}
	return true
}

func findActiveMarketByTime(prefix string, nowET time.Time, searchHours, stepHours int) (map[string]any, string, error) {
	if prefix == "" {
		return nil, "", fmt.Errorf("missing prefix")
	}
	base := time.Date(nowET.Year(), nowET.Month(), nowET.Day(), nowET.Hour(), 0, 0, 0, nowET.Location())
	nowUTC := nowET.UTC()

	offsets := []int{0}
	for h := stepHours; h <= searchHours; h += stepHours {
		offsets = append(offsets, h, -h)
	}

	var fallbackMarket map[string]any
	var fallbackSlug string
	for _, offset := range offsets {
		candidate := base.Add(time.Duration(offset) * time.Hour)
		slug := buildSlug(prefix, candidate)
		market, err := fetchMarketBySlug(slug)
		if err != nil {
			continue
		}
		if enable, ok := market["enableOrderBook"].(bool); ok && !enable {
			if fallbackMarket == nil {
				fallbackMarket = market
				fallbackSlug = slug
			}
			continue
		}
		if isOpenMarket(market, nowUTC) {
			return market, slug, nil
		}
		if fallbackMarket == nil {
			fallbackMarket = market
			fallbackSlug = slug
		}
	}

	if fallbackMarket != nil {
		return fallbackMarket, fallbackSlug, nil
	}
	return nil, "", fmt.Errorf("no market found for prefix: %s", prefix)
}

func resolveYesNoTokens(market map[string]any, slug string) (MarketTokens, error) {
	outcomes := normalizeListField(market["outcomes"])
	coerced := make([]string, 0, len(outcomes))
	for _, o := range outcomes {
		coerced = append(coerced, coerceOutcome(o))
	}
	tokenIDs := extractClobTokenIDs(market)
	if len(coerced) == 0 || len(tokenIDs) == 0 {
		return MarketTokens{}, fmt.Errorf("missing outcomes or clobTokenIds")
	}

	yesLike := map[string]bool{"yes": true, "true": true, "up": true}
	noLike := map[string]bool{"no": true, "false": true, "down": true}
	mapped := map[string]string{}
	for i, outcome := range coerced {
		norm := strings.ToLower(strings.TrimSpace(outcome))
		if yesLike[norm] {
			mapped["yes"] = tokenIDs[i]
		} else if noLike[norm] {
			mapped["no"] = tokenIDs[i]
		}
	}
	if mapped["yes"] == "" || mapped["no"] == "" {
		if len(tokenIDs) == 2 {
			if mapped["yes"] == "" {
				mapped["yes"] = tokenIDs[0]
			}
			if mapped["no"] == "" {
				mapped["no"] = tokenIDs[1]
			}
		} else {
			return MarketTokens{}, fmt.Errorf("could not map outcomes to yes/no")
		}
	}

	return MarketTokens{
		YesTokenID:      mapped["yes"],
		NoTokenID:       mapped["no"],
		Outcomes:        coerced,
		ClobTokenIDs:    tokenIDs,
		MarketID:        toString(market["id"]),
		Slug:            fallbackString(toString(market["slug"]), slug),
		EnableOrderbook: toBoolPtr(market["enableOrderBook"]),
		Closed:          toBoolPtr(market["closed"]),
		Active:          toBoolPtr(market["active"]),
		StartDate:       toString(market["startDate"]),
		EndDate:         toString(market["endDate"]),
	}, nil
}

func extractClobTokenIDs(market map[string]any) []string {
	items := normalizeListField(market["clobTokenIds"])
	out := make([]string, 0, len(items))
	for _, item := range items {
		if s := strings.TrimSpace(toString(item)); s != "" {
			out = append(out, s)
		}
	}
	return out
}

func resolveMarketTokens(cfg TraderConfig, nowET time.Time) (MarketTokens, error) {
	slugValue := ""
	if cfg.Slug != "" {
		slugValue = normalizeSlug(cfg.Slug)
	}
	if cfg.TokenIDUp != "" && cfg.TokenIDDown != "" {
		return MarketTokens{
			YesTokenID: cfg.TokenIDUp,
			NoTokenID:  cfg.TokenIDDown,
			Slug:       slugValue,
		}, nil
	}

	autoSlug := cfg.AutoSlug
	if autoSlug {
		prefix := cfg.SlugPrefix
		if prefix == "" {
			prefix = inferSlugPrefix(slugValue)
		}
		if prefix == "" {
			return MarketTokens{}, fmt.Errorf("provide --slug-prefix or a slug with time suffix")
		}
		market, slug, err := findActiveMarketByTime(prefix, nowET, cfg.SearchHours, cfg.SearchStepHours)
		if err != nil {
			return MarketTokens{}, err
		}
		return resolveYesNoTokens(market, slug)
	}

	if slugValue == "" {
		return MarketTokens{}, fmt.Errorf("provide --slug or enable --auto-slug with --slug-prefix")
	}
	market, err := fetchMarketBySlug(slugValue)
	if err != nil {
		return MarketTokens{}, err
	}
	return resolveYesNoTokens(market, slugValue)
}

func normalizeListField(value any) []any {
	if value == nil {
		return nil
	}
	switch v := value.(type) {
	case []any:
		return v
	case []string:
		out := make([]any, 0, len(v))
		for _, s := range v {
			out = append(out, s)
		}
		return out
	case string:
		text := strings.TrimSpace(v)
		if text == "" {
			return nil
		}
		var parsed any
		if err := json.Unmarshal([]byte(text), &parsed); err == nil {
			if items, ok := parsed.([]any); ok {
				return items
			}
		}
		if strings.Contains(text, ",") {
			parts := strings.Split(text, ",")
			out := make([]any, 0, len(parts))
			for _, part := range parts {
				p := strings.TrimSpace(part)
				if p != "" {
					out = append(out, p)
				}
			}
			return out
		}
		return []any{text}
	default:
		return []any{v}
	}
}

func coerceOutcome(value any) string {
	if value == nil {
		return ""
	}
	if row, ok := value.(map[string]any); ok {
		for _, key := range []string{"title", "name", "label", "outcome"} {
			if raw, ok := row[key]; ok {
				return toString(raw)
			}
		}
	}
	return toString(value)
}

func toBoolPtr(value any) *bool {
	if value == nil {
		return nil
	}
	if b, ok := value.(bool); ok {
		return &b
	}
	return nil
}

func fallbackString(value, fallback string) string {
	if value != "" {
		return value
	}
	return fallback
}

func toString(value any) string {
	if value == nil {
		return ""
	}
	switch v := value.(type) {
	case string:
		return v
	case fmt.Stringer:
		return v.String()
	case []byte:
		return string(v)
	default:
		return fmt.Sprintf("%v", v)
	}
}

func toSlice(value any) []any {
	if value == nil {
		return nil
	}
	if v, ok := value.([]any); ok {
		return v
	}
	return nil
}

File: backtest/libs/polymarket_utils.py
from dataclasses import dataclass
import datetime as dt
import json
import re
from typing import List, Optional, Tuple
import urllib.parse
import urllib.request
from zoneinfo import ZoneInfo

GAMMA_MARKETS_BY_SLUG = "https://gamma-api.polymarket.com/markets?slug="
GAMMA_UA = "Mozilla/5.0 (compatible; CodexBot/1.0)"
ET_TZ = ZoneInfo("America/New_York")
MONTH_NAMES = [
    "",
    "january",
    "february",
    "march",
    "april",
    "may",
    "june",
    "july",
    "august",
    "september",
    "october",
    "november",
    "december",
]
SLUG_TIME_RE = re.compile(
    r"^(?P<prefix>.+)-"
    r"(?P<month>january|february|march|april|may|june|july|august|september|october|november|december)-"
    r"(?P<day>\d{1,2})-"
    r"(?P<hour>\d{1,2})(?P<ampm>am|pm)-et$"
)


@dataclass
class MarketTokens:
    yes_token_id: str
    no_token_id: str
    outcomes: List[str]
    clob_token_ids: List[str]
    market_id: str
    slug: str
    enable_orderbook: Optional[bool]
    closed: Optional[bool]
    active: Optional[bool]
    start_date: Optional[str]
    end_date: Optional[str]


def normalize_slug(value: str) -> str:
    marker = "polymarket.com/event/"
    if marker in value:
        return value.split(marker, 1)[1].strip("/")
    return value


def fetch_market_by_slug(slug: str) -> dict:
    slug = normalize_slug(slug)
    url = GAMMA_MARKETS_BY_SLUG + urllib.parse.quote(slug)
    req = urllib.request.Request(url, headers={"User-Agent": GAMMA_UA})
    with urllib.request.urlopen(req) as resp:
        data = json.load(resp)

    if isinstance(data, dict):
        markets = data.get("markets", [])
    else:
        markets = data

    if not markets:
        raise ValueError(f"No market found for slug: {slug}")
    return markets[0]


def infer_slug_prefix(slug: str) -> Optional[str]:
    match = SLUG_TIME_RE.match(slug)
    if not match:
        return None
    return match.group("prefix")


def build_slug(prefix: str, when_et: dt.datetime) -> str:
    month = MONTH_NAMES[when_et.month]
    day = when_et.day
    hour24 = when_et.hour
    hour12 = hour24 % 12
    if hour12 == 0:
        hour12 = 12
    ampm = "am" if hour24 < 12 else "pm"
    return f"{prefix}-{month}-{day}-{hour12}{ampm}-et"


def _parse_iso_dt(value: Optional[str]) -> Optional[dt.datetime]:
    if not value:
        return None
    try:
        if value.endswith("Z"):
            value = value[:-1] + "+00:00"
        return dt.datetime.fromisoformat(value)
    except ValueError:
        return None


def _is_open_market(market: dict, now_utc: dt.datetime) -> bool:
    closed = market.get("closed")
    if closed is True:
        return False
    start_dt = _parse_iso_dt(market.get("startDate"))
    if start_dt and now_utc < start_dt:
        return False
    end_dt = _parse_iso_dt(market.get("endDate"))
    if end_dt and end_dt < now_utc:
        return False
    return True


def find_active_market_by_time(
    prefix: str,
    now_et: Optional[dt.datetime] = None,
    search_hours: int = 6,
    step_hours: int = 1,
) -> Tuple[dict, str]:
    if now_et is None:
        now_et = dt.datetime.now(tz=ET_TZ)
    base = now_et.replace(minute=0, second=0, microsecond=0)
    now_utc = now_et.astimezone(dt.timezone.utc)

    offsets = [0]
    for h in range(1, search_hours + 1, step_hours):
        offsets.append(h)
        offsets.append(-h)

    fallback = None
    for offset in offsets:
        candidate = base + dt.timedelta(hours=offset)
        slug = build_slug(prefix, candidate)
        try:
            market = fetch_market_by_slug(slug)
        except Exception:
            continue
        if market.get("enableOrderBook") is False:
            fallback = fallback or (market, slug)
            continue
        if _is_open_market(market, now_utc):
            return market, slug
        fallback = fallback or (market, slug)

    if fallback:
        return fallback
    raise ValueError(f"No market found for prefix: {prefix}")


def resolve_yes_no_tokens(market: dict, slug: str) -> MarketTokens:
    outcomes = [_coerce_outcome(o) for o in _normalize_list_field(market.get("outcomes"))]
    token_ids = extract_clob_token_ids(market)
    if not outcomes or not token_ids:
        raise ValueError("Missing outcomes or clobTokenIds in market response.")

    yes_like = {"yes", "true", "up"}
    no_like = {"no", "false", "down"}

    mapped = {}
    for outcome, token_id in zip(outcomes, token_ids):
        norm = outcome.strip().lower()
        if norm in yes_like:
            mapped["yes"] = token_id
        elif norm in no_like:
            mapped["no"] = token_id

    if "yes" not in mapped or "no" not in mapped:
        if len(token_ids) == 2:
            mapped.setdefault("yes", token_ids[0])
            mapped.setdefault("no", token_ids[1])
        else:
            raise ValueError(
                "Could not map outcomes to yes/no. Provide explicit token IDs."
            )

    return MarketTokens(
        yes_token_id=mapped["yes"],
        no_token_id=mapped["no"],
        outcomes=outcomes,
        clob_token_ids=token_ids,
        market_id=str(market.get("id", "")),
        slug=str(market.get("slug", slug)),
        enable_orderbook=market.get("enableOrderBook"),
        closed=market.get("closed"),
        active=market.get("active"),
        start_date=market.get("startDate"),
        end_date=market.get("endDate"),
    )


def extract_clob_token_ids(market: dict) -> List[str]:
    token_ids = _normalize_list_field(market.get("clobTokenIds"))
    return [str(t) for t in token_ids if t]


def _coerce_outcome(value) -> str:
    if isinstance(value, dict):
        for key in ("title", "name", "label", "outcome"):
            if key in value:
                return str(value[key])
    return str(value)


def _normalize_list_field(value) -> List:
    if value is None:
        return []
    if isinstance(value, list):
        return value
    if isinstance(value, tuple):
        return list(value)
    if isinstance(value, str):
        text = value.strip()
        if not text:
            return []
        try:
            parsed = json.loads(text)
        except json.JSONDecodeError:
            parsed = None
        if isinstance(parsed, list):
            return parsed
        if "," in text:
            return [chunk.strip() for chunk in text.split(",") if chunk.strip()]
        return [text]
    return [value]
